File: scripts/aux/extractSingleTree.pl
#!/usr/bin/env perl
use strict;
use warnings;
use PDL;
use PDL::NiceSlice;
use PDL::IO::HDF5;
use Data::Dumper;
$PDL::BIGPDL = 1;

# Extract a single merger tree from a merger tree file. Useful for creation of test cases.
# Andrew Benson (19-September-2012)

# Get arguments.
die("Usage: extractSingleTree.pl <fromFile> <toFile> <forestIndex>")
    unless ( scalar(@ARGV) == 3 );
my $inFileName  = $ARGV[0];
my $outFileName = $ARGV[1];
my $tree        = $ARGV[2];

# Remove the output file.
unlink($outFileName);

# Open the files.
my $inFile  = new PDL::IO::HDF5(     $inFileName);
my $outFile = new PDL::IO::HDF5(">".$outFileName);

# Detect file version and set group names appropriately.
(my $inFormatVersion) = (grep {$_ eq "formatVersion"} $inFile->attrs()) ? $inFile->attrGet("formatVersion") : (1);
my $forestName      = $inFormatVersion == 1 ? "treeIndex" : "forestIndex";
my $halosName       = $inFormatVersion == 1 ? "haloTrees" : "forestHalos";

# Find the tree to extract.
my $forestIndex = $inFile->group($forestName)->dataset($forestName    )->get();
my $firstNode   = $inFile->group($forestName)->dataset("firstNode"    )->get();
my $nodeCount   = $inFile->group($forestName)->dataset("numberOfNodes")->get();
my $selected    = which($forestIndex == $tree);
my $start       = $firstNode->index($selected)->sclr();
my $count       = $nodeCount->index($selected)->sclr();
my $end         = $start+$count-1;

# Read particle datasets if present.
my $particles;
my $particlesExtracted;
if ( grep {$_ eq "particles"} $inFile->groups() ) {
    my $particleGroup           = $inFile       ->group   ('particles')                                      ;
    my @particleNames           = $particleGroup->datasets(           )                                      ;
    my $particleIndexCount      = $inFile       ->group   ($halosName )->dataset("particleIndexCount")->get();
    my $particleIndexStart      = $inFile       ->group   ($halosName )->dataset("particleIndexStart")->get();
    # Create copies of all datasets in the "particles" group sized to hold just those particles from our tree-to-be-extracted.
    my $present                 = which($particleIndexCount->($start:$end) > 0);
    my $particlesExtractedTotal = $particleIndexCount->($start:$end)->($present)->sum();
    foreach my $datasetName ( @particleNames ) {
	$particles->{$datasetName} = $particleGroup->dataset($datasetName)->get();
	my $dimensions = $particles->{$datasetName}->ndims();
	if ( $particles->{$datasetName}->type() eq "longlong" ) {	    
	    if ( $dimensions == 1 ) {
		$particlesExtracted->{$datasetName} = pdl longlong zeros(                                      $particlesExtractedTotal);
	    } elsif ( $dimensions == 2 ) {
		$particlesExtracted->{$datasetName} = pdl longlong zeros($particles->{$datasetName}->getdim(0),$particlesExtractedTotal);
	    } else {
		die("extractSingleTree.pl: unable to handle rank > 2 datasets");
	    }
	} else {
	    if ( $dimensions == 1 ) {
		$particlesExtracted->{$datasetName} = pdl          zeros(                                      $particlesExtractedTotal);
	    } elsif ( $dimensions == 2 ) {
		$particlesExtracted->{$datasetName} = pdl          zeros($particles->{$datasetName}->getdim(0),$particlesExtractedTotal);
	    } else {
		die("extractSingleTree.pl: unable to handle rank > 2 datasets");
	    }
	}
    }
    # Copy in particles used by the tree to be extracted to our new arrays, and recompute the indices into these new arrays,
    my $particleI = 0;
    for(my $i=$start;$i<=$end;++$i) {
	my $particleStart = $particleIndexStart->(($i))->sclr();
	my $particleCount = $particleIndexCount->(($i))->sclr();
	next
	    if ( $particleCount <= 0 );
	my $particleEnd   = $particleStart+$particleCount-1;
	foreach my $datasetName ( @particleNames ) {
	    my $dimensions = $particles->{$datasetName}->ndims();
	    if ( $dimensions == 1 ) {
		$particlesExtracted->{$datasetName}->(  $particleI:$particleI+$particleCount-1) .= $particles->{$datasetName}->(  $particleStart:$particleEnd);
	    } elsif ( $dimensions == 2 ) {
		$particlesExtracted->{$datasetName}->(:,$particleI:$particleI+$particleCount-1) .= $particles->{$datasetName}->(:,$particleStart:$particleEnd);
	    } else {
		die("extractSingleTree.pl: unable to handle rank > 2 datasets");
	    }
	}
	$particleIndexStart->(($i)) .= $particleI;
	$particleI                  += $particleCount;
    }
    $outFile->group("forestHalos")->dataset("particleIndexStart")->set($particleIndexStart->($start:$end));
    my $particlesOut = $outFile->group("particles");
    foreach my $datasetName ( @particleNames ) {
	$particlesOut->dataset($datasetName)->set($particlesExtracted->{$datasetName});
    }
}

# Read all halo datasets.
foreach my $datasetName ( $inFile->group($halosName)->datasets() ) {
    # Skip the particleIndexStart dataset as we have already recomputed and output this array above.
    next
	if ( $datasetName eq "particleIndexStart" );
    my $dataset    = $inFile->group($halosName)->dataset($datasetName)->get();
    my $dimensions = $dataset->ndims();
     if      ( $dimensions == 1 ) {
     	$outFile->group("forestHalos")->dataset($datasetName)->set($dataset->(  $start:$end));
     } elsif ( $dimensions == 2 ) {
     	$outFile->group("forestHalos")->dataset($datasetName)->set($dataset->(:,$start:$end));
     } else {
     	die("extractSingleTree.pl: unable to handle rank > 2 datasets");
     }
}

# Create the forestIndex group.
$outFile->group("forestIndex")->dataset("forestIndex"  )->set(pdl longlong([$tree ]));
$outFile->group("forestIndex")->dataset("firstNode"    )->set(pdl longlong([     0]));
$outFile->group("forestIndex")->dataset("numberOfNodes")->set(pdl longlong([$count]));

# Set format version.
my $formatVersion;
if ( grep {$_ eq "formatVersion"} $outFile->attrs() ) {
    $formatVersion = $inFile->attrGet("formatVersion");
} else {
    $formatVersion = pdl long 2;
}
$outFile->attrSet("formatVersion" => $formatVersion);

# Copy all attributes.
foreach my $groupName ( $inFile->groups() ) {
    my $outGroupName = $groupName;
    $outGroupName = "forestIndex"
	if ( $groupName eq "treeIndex" );
    $outGroupName = "forestHalos"
	if ( $groupName eq "haloTrees" );
    next 
	if ( $groupName eq "particles" );
    my $fromGroup = $inFile ->group(   $groupName);
    my $toGroup   = $outFile->group($outGroupName);
    &Copy_Attributes($fromGroup,$toGroup);
}

exit;

sub Copy_Attributes {
    # Copy all attributes from an object in the input files to the output file.
    # Get the objects from and to.
    my $objectFrom = shift;
    my $objectTo   = shift;
  
    # Copy all attributes.
    my @attributes = $objectFrom->attrs();
    foreach my $attribute ( @attributes ) {
	my @attrValue = $objectFrom->attrGet($attribute);
	$objectTo->attrSet($attribute => $attrValue[0]);
    }
}

File: scripts/aux/extractSingleTree.py
#!/usr/bin/env python3
import os
import sys
import numpy as np
import h5py
import argparse

# Extract a single merger tree from a merger tree file. Useful for creation of test cases.
# Andrew Benson (19-September-2012; ported to Python 23-August-2024)

# Convert command line arguments to integers.
def restricted_integer(x):
    try:
        x = int(x)
    except ValueError:
        raise argparse.ArgumentTypeError("%r not a integer literal" % (x,))
    return x

# Parse command line arguments.
parser = argparse.ArgumentParser(prog='extractSingleTree.py',description='Extract a single forest from a Galacticus merger tree file to its own file')
parser.add_argument('fromFile'                           )
parser.add_argument('toFile'                             )
parser.add_argument('forestIndex',type=restricted_integer)
args = parser.parse_args()

# Remove the output file.
if os.path.exists(args.toFile):
    os.remove(args.toFile)

# Open the files.
fromFile = h5py.File(args.fromFile,'r')
toFile   = h5py.File(args.  toFile,'w')

# Detect file version and set group names appropriately.
fromFormatVersion = fromFile.attrs['formatVersion'] if "formatVersion" in fromFile.attrs else 1
forestName      = "treeIndex" if fromFormatVersion == 1 else "forestIndex"
halosName       = "haloTrees" if fromFormatVersion == 1 else "forestHalos"

# Find the forest to extract.
print("Locating forest...")
forestIndex = fromFile[forestName][forestName     ][:]
firstNode   = fromFile[forestName]['firstNode'    ][:]
nodeCount   = fromFile[forestName]['numberOfNodes'][:]
selected    = forestIndex == args.forestIndex
if len(np.nonzero(selected)) < 1:
    sys.exit("failed to find request forest")
if len(np.nonzero(selected)) > 1:
    sys.exit("found multiple matching forests")
start       = firstNode[np.nonzero(selected)]
count       = nodeCount[np.nonzero(selected)]
end         = start+count
print("...done")

# Read all halo datasets.
print("Extracting datasets...")
toFile.create_group("forestHalos")
for name, h5obj in fromFile[halosName].items():
    # Ignore groups.
    if not isinstance(h5obj,h5py.Dataset):
        continue
    # Skip the particleIndex datasets as we will need to recompute them later.
    if name == "particleIndexStart" or name == "particleIndexCount":
        continue
    print("   "+name)
    dataset    = fromFile[halosName][name][start[0]:end[0],...]
    toFile["forestHalos"].create_dataset(name,data=dataset)
print("...done")

# Create the forestIndex group.
toFile.create_group("forestIndex")
toFile["forestIndex"].create_dataset("forestIndex"  ,data=np.array([args.forestIndex]))
toFile["forestIndex"].create_dataset("firstNode"    ,data=np.array([               0]))
toFile["forestIndex"].create_dataset("numberOfNodes",data=                     count  )

# Copy the particle data if present.
if "particleIndexStart" in fromFile[halosName]:
    print("Extracting particle data...")
    particleIndexStart = fromFile[halosName]['particleIndexStart'][start[0]:end[0]]
    particleIndexCount = fromFile[halosName]['particleIndexCount'][start[0]:end[0]]
    indices = set()
    for i in range(len(particleIndexStart)):
        if particleIndexStart[i] < 0:
            continue
        indices.update(list(range(particleIndexStart[i],particleIndexStart[i]+particleIndexCount[i]+1)))
    indices = np.sort(np.array(list(indices)))
    toFile.create_group("particles")
    for name, h5obj in fromFile['particles'].items():
        # Ignore groups.
        if not isinstance(h5obj,h5py.Dataset):
            continue
        dataset = fromFile['particles'][name][indices]
        toFile["particles"].create_dataset(name,data=dataset)
    particleIndexStartNew = particleIndexStart
    for i in range(len(particleIndexStartNew)):
        if particleIndexStartNew[i] < 0:
            continue
        particleIndexStartNew[i] = np.nonzero(indices == particleIndexStart[i])[0][0]
    toFile[halosName].create_dataset("particleIndexStart",data=particleIndexStartNew)
    toFile[halosName].create_dataset("particleIndexCount",data=particleIndexCount   )
    print("...done")
    
# Set format version.
toFile.attrs["formatVersion"] = 2

# Copy all attributes.
print("Copying attributes....")
for name, h5obj in fromFile.items():
    # Ignore non-groups.
    if not isinstance(h5obj,h5py.Group):
        continue
    # Translate names between versions.
    nameTo = name
    if name == "treeIndex":
        nameTo = "forestIndex"
    if name == "haloTrees":
        nameTo = "forestHalos"
    # Copy the attributes.
    print("   "+name+" -> "+nameTo)
    if not nameTo in toFile:
        toFile.create_group(nameTo)
    toFile[nameTo].attrs.update(h5obj.attrs)
print("...done")
